src/main/java/org/spongepowered/api/data/generator/DataGenerator.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.data.generator;
+
+import org.spongepowered.api.data.DataRegistration;
+import org.spongepowered.api.data.key.Key;
+import org.spongepowered.api.data.manipulator.DataManipulator;
+import org.spongepowered.api.data.manipulator.ImmutableDataManipulator;
+import org.spongepowered.api.plugin.PluginContainer;
+import org.spongepowered.api.util.ResettableBuilder;
+
+/**
+ * A generator that allows a {@link DataRegistration} to be constructed with
+ * generated {@link DataManipulator} and {@link ImmutableDataManipulator} classes.
+ */
+public interface DataGenerator<
+        M extends DataManipulator<M, I>,
+        I extends ImmutableDataManipulator<I, M>,
+        G extends DataGenerator<M, I, G, R>,
+        R extends DataGenerator<?, ?, ?, R>>
+        extends ResettableBuilder<DataRegistration<?,?>, R> {
```

Comment:
Like specifically why the `R` random type.

Reply:
This one is used when resetting the builder, to reset the generic types as well.